src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+using Microsoft.Win32.SafeHandles;
+
+using Internal.Cryptography;
+
+using ErrorCode = Interop.NCrypt.ErrorCode;
+using AsymmetricPaddingMode = Interop.NCrypt.AsymmetricPaddingMode;
+using BCRYPT_OAEP_PADDING_INFO = Interop.BCrypt.BCRYPT_OAEP_PADDING_INFO;
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class RSACng : RSA
+    {
+        /// <summary>
+        ///     Encrypts data using the public key.
+        /// </summary>
+        public override byte[] Encrypt(byte[] data, RSAEncryptionPadding padding)
+        {
+            unsafe
+            {
+                return EncryptOrDecrypt(data, padding, Interop.NCrypt.NCryptEncrypt);
+            }
+        }
+
+        /// <summary>
+        ///     Decrypts data using the private key.
+        /// </summary>
+        public override byte[] Decrypt(byte[] data, RSAEncryptionPadding padding)
+        {
+            unsafe
+            {
+                return EncryptOrDecrypt(data, padding, Interop.NCrypt.NCryptDecrypt);
+            }
+        }
+
+        //
+        // Conveniently, Encrypt() and Decrypt() are identical save for the actual P/Invoke call to CNG. Thus, both
+        // api's invoke this common helper with the "transform" parameter determining whether encryption or decryption is done.
+        //
+        private byte[] EncryptOrDecrypt(byte[] data, RSAEncryptionPadding padding, EncryptOrDecryptAction encryptOrDecrypt)
+        {
+            if (data == null)
+                throw new ArgumentNullException("data");
+
+            if (padding == null)
+                throw new ArgumentNullException("padding");
+
+            unsafe
```

Comment:
There are two call-sites to EncryptOrDecrypt (L60 and L74), and more if we add support for new padding modes.



From: Stephen Toub [mailto:notifications@github.com]
Sent: Thursday, August 6, 2015 12:14 PM
To: dotnet/corefx
Cc: Atsushi Kanamori
Subject: Re: [corefx] Implement System.Security.Cryptography.RSACng (#2654)


In src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs<https://github.com/dotnet/corefx/pull/2654#discussion_r36453587>:

> +            }

> +        }

> +

> +        //

> +        // Conveniently, Encrypt() and Decrypt() are identical save for the actual P/Invoke call to CNG. Thus, both

> +        // api's invoke this common helper with the "transform" parameter determining whether encryption or decryption is done.

> +        //

> +        private byte[] EncryptOrDecrypt(byte[] data, RSAEncryptionPadding padding, EncryptOrDecryptAction encryptOrDecrypt)

> +        {

> +            if (data == null)

> +                throw new ArgumentNullException("data");

> +

> +            if (padding == null)

> +                throw new ArgumentNullException("padding");

> +

> +            unsafe

It looks like the only place that needs this is the call to EncryptOrDecrypt (L74). If that's true, consider wrapping just that line in unsafe.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/2654/files#r36453587>.